### ISSUE TITLE
disable writing to nx cache

### ISSUE THREAD

**Issue by cdaringe** [nrwl/nx#16182]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description

I want to _completely_ disable all NX caching capability in some circumstances.

## Motivation

Some tasks are purely ephemeral and known to not need caching. Caching can be slow in some circumstances, purely by nature of i/o.

## Suggested Implementation

https://nx.dev/reference/environment-variables

`NX_SKIP_NX_CACHE` should fully opt out of _all_ caching, not just read caching.

`NX_SKIP_NX_CACHE_READ` and `NX_SKIP_NX_CACHE_WRITE` could be used for finer grain tuning.

In the compiled source, we can see a flag `doNotSkipCache`, which is **only** considered for reads, not the writes:
```js
    applyFromCacheOrRunTask(doNotSkipCache, task, groupId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.preRunSteps([task], { groupId });
            // hash the task here
            let results = doNotSkipCache ? yield this.applyCachedResults([task]) : [];
            // the task wasn't cached
            if (results.length === 0) {
                // cache prep
                const taskOutputs = (0, utils_1.getOutputs)(this.projectGraph.nodes, task);
                yield this.cache.removeRecordedOutputsHashes(taskOutputs);
                const { code, terminalOutput } = yield this.runTaskInForkedProcess(task);
                results.push({
                    task,
                    status: code === 0 ? 'success' : 'failure',
                    terminalOutput,
                });
            }
            yield this.postRunSteps(results, { groupId });
        });
    }
```

Further, cache writing is split between this fn ^ above, _and_ `postRunSteps`. They ought be considered for consolidation

## Alternate Implementations

- CLI params to set the value
- Cache instance vs FakeCache instance, where the later no-ops all of the things (could be problematic for some cases, as the cache instance hosts all sorts of non i/o methods, like providing cache paths to other entities)

**Comment by slaesh** [comment]
we are currently experiencing something similar.
would be great to have it finer..
`skip-reading-remote-cache` and `skip-writing-remote-cache` ..

I guess it is only effecting the remote cache? or is it effecting the local cache too?

our use-case is.. from time to time we want to just build everything.. regardless of any cache..
since we observed some issues from time to time that the cached versions are not that compatible..
so we would like to skip the remote cache (only read), but want to write it afterwards since we want to use the benefits of doing the hard-work only once! :)

even though the docs arent clear what the `skip-nx-cahce` is doing..

it says

```
NX_SKIP_NX_CACHE | boolean | Rerun the tasks even when the results are available in the cache
```

so.. does it write or not? you could more likely read it as "does not read, but will write" ..

but if you view the CLI output, it says

```
>  NX   --skip-nx-cache disables the connection to Nx Cloud for the current run.

The remote cache will not be read from or written to during this run.
```

**Comment by adolfov** [comment]
With the NX Daemon enabled it seems the cache is always active.
The parameter is explicitly set to "true".
https://github.com/nrwl/nx/blob/master/packages/nx/src/daemon/server/project-graph-incremental-recomputation.ts#L266C11-L266C11

If the NX Daemon is disabled, an env var is used to check whether the cache will be written.
https://github.com/nrwl/nx/blob/master/packages/nx/src/project-graph/project-graph.ts#L79
Env var is: NX_CACHE_PROJECT_GRAPH

This worked for me to disable cache write:
```
ENV NX_DAEMON=false
ENV NX_SKIP_NX_CACHE=true
ENV NX_CACHE_PROJECT_GRAPH=false
```
_Thanks to [Patrick](https://nrwlcommunity.slack.com/archives/CMFKWPU6Q/p1688676159591679?thread_ts=1688659894.080609&cid=CMFKWPU6Q) in the slack community_

**Comment by vid** [comment]
I second this. There are conditions where it doesn't work as expected, and it just adds a layer of confusing indirection where the benefit isn't worth it. 

I now have a workflow where I completely remove anything cache related before building, but I still end up with this after an edit:

```
$ for i in `find -name AStorage.js`; do echo $i; grep oox $i; done  
./node_modules/.cache/nx/14744362427699911472/outputs/modules/domain-storage/build/AStorage.js
            console.log('\nooxi.', e.name, filter, filter === undefined, e.name.match(filter)?.length);
./node_modules/.cache/nx/6607176757028174651/outputs/modules/domain-storage/build/AStorage.js
            console.log('\nooxis.', e.name, filter, filter === undefined, e.name.match(filter)?.length);
./modules/domain-storage/build/AStorage.js
            console.log('\nooxis.', e.name, filter, filter === undefined, e.name.match(filter)?.length);
```

If I use `NX_CACHE_PROJECT_GRAPH`, the whole project breaks.

I really want to spend -zero- time debugging what my build tools are thinking during dev.

If I use

**Comment by FrozenPandaz** [comment]
I think the original request of this issue has been fulfilled. `--skipNxCache` now disables both reading and writing to the cache.

The `NX_CACHE_PROJECT_GRAPH` is a separate in the Daemon is a separate issue. Please open a new issue for that.

**Comment by rlmestre** [comment]
@FrozenPandaz since what version is this implemented? Thanks
